### PR TITLE
[SYCL][COMPAT] Added kernel.hpp and tests

### PR DESCRIPTION
**Pull request by Alcpz**
SYCLcompat primary goals and features are documented [here](https://intel.github.io/llvm-docs/syclcompat/README.html).

The PR includes the helper code to assist with kernel source-to-source translation. 
This PR only includes tests for linux environments .

### CODE REVIEW

**Review by AidanBeltonS — COMMENTED** [review]
Other than the liscense text LGTM

File: sycl/test-e2e/syclcompat/kernel/kernel_function_lin.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+// ===---------------------------------------------------------------------===//
+
+// REQUIRES: linux
+
+// RUN: %clangxx -fPIC -shared -fsycl -fsycl-targets=%{sycl_triple} %S/Inputs/kernel_module_lin.cpp -o %t.so
+// RUN: %clangxx -D__TEST_SHARED_LIB='"%t.so"' -ldl -fsycl -fsycl-targets=%{sycl_triple} %t.so %s -o %t.out
```
Comment (joeatodd) [inline sycl/test-e2e/syclcompat/kernel/kernel_function_lin.cpp:36]: Identifiers beginning with double underscore are reserved unconditionally.

**Review by joeatodd — CHANGES_REQUESTED** [review]
LGTM aside from reserved name.

File: sycl/test-e2e/syclcompat/kernel/kernel_function_lin.cpp (hunk 1 above)
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/kernel/kernel_function_lin.cpp]: Addressed in last commit

**Review by joeatodd — APPROVED** [review]

**Comment by Alcpz** [comment]
Hello, @intel/llvm-gatekeepers we think this PR is ready to be merged. Thank you!

**Pull request merged**